test/unit/BPool/BPool_JoinPool.t.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {BPoolBase} from './BPoolBase.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import {BNum} from 'contracts/BNum.sol';
+import {IBPool} from 'interfaces/IBPool.sol';
+
+contract BPoolJoinPool is BPoolBase {
+  uint256 public poolAmountOut = 10e18;
+  // Assets Under Management
+  uint256 public tokenAUM = 10e18;
+  uint256 public secondTokenAUM = 30e18;
+
+  uint256 public requiredTokenIn = 1e18;
+  uint256 public requiredSecondTokenIn = 3e18;
+
+  function setUp() public virtual override {
+    super.setUp();
+    bPool.set__finalized(true);
+    // simulate expected finalize outcome
+    bPool.call__mintPoolShare(INIT_POOL_SUPPLY);
+    bPool.call__pushPoolShare(deployer, INIT_POOL_SUPPLY);
+    address[] memory _tokens = new address[](2);
+    _tokens[0] = token;
+    _tokens[1] = secondToken;
+    bPool.set__tokens(_tokens);
+    // token weights are not used for all-token joins
+    _setRecord(token, IBPool.Record({bound: true, index: 0, denorm: 0}));
+    _setRecord(secondToken, IBPool.Record({bound: true, index: 1, denorm: 0}));
+    // underlying balances are used instead
+    vm.mockCall(token, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(tokenAUM)));
+    vm.mockCall(secondToken, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(secondTokenAUM)));
+  }
+
+  function test_RevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolIsNotFinalized() external {
+    bPool.set__finalized(false);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_PoolNotFinalized.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  // should not happen in the real world since finalization mints 100 tokens
+  // and sends them to controller
+  function test_RevertWhen_TotalSupplyIsZero() external {
+    // undo what was just done by setup
+    bPool.call__pullPoolShare(deployer, INIT_POOL_SUPPLY);
+    bPool.call__burnPoolShare(INIT_POOL_SUPPLY);
+    // it should revert
+    vm.expectRevert(BNum.BNum_DivZero.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolAmountOutIsZero() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidPoolRatio.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolRatioRoundsDownToZero() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidPoolRatio.selector);
+    // somewhat coupled to BNum, perhaps not a different case from the above?
+    bPool.joinPool(49, new uint256[](2));
+  }
```

Comment:
i think these 2 could be merged, if you feel like, fuzzing amountOut in between 0 and `totalSupply/2-1` should do the trick i think